src/Interpreters/GraceHashJoin.h

```diff
@@ -139,6 +139,8 @@ class GraceHashJoin final : public IJoin
     mutable SharedMutex rehash_mutex;
 
     FileBucket * current_bucket = nullptr;
+
+    /// Function `getDelayedBlocks` is a critical section, we process only one bucket at a time.
```

Comment:
This comment looks misleading. While only one `getDelayedBlocks()` should be executed at a time, it's not because of the mutex, but because of the pipeline is built for grace join. So, probably, better remove it.

In general, this mutex looks redundant. But this matter can be addressed separately to allow now land the fix.